accepted/single-file/design.md

```diff
@@ -0,0 +1,281 @@
+{
+    public partial class Bundle
+    {
+        public static System.IO.Stream GetFileStream(Assembly assembly, string name);
+    }
+}
+```
+
+We can also provide an abstraction that abstracts away the physical location of a file (bundle or disk). For example, add a variant of  `GetFileStream` API that looks for a file in the bundle, and if not found, falls back to disk-lookup.
+
+#### Single-file Plugins
+
+The above design should be extended to seamlessly support single-file publish for plugins.
+
+- The bundler tool will mostly work as-is, regardless of whether we publish an application or class-lib. The binary blob with dependencies can be appended to both native and managed binaries.
+- For host/runtime support, the options are:
+  - Implement plugins using existing infrastructure. For example: Take control of assembly/native binary loads via existing `AssemblyLoadContext` callbacks and events. Extract the files embedded within the single-file plugin using the `GetFileStream()` API and load them on demand.
+  - Have new API to load a single-file plugin, for example: `AssemblyLoadContext.LoadWithEmbeddedDependencies()`.
+
+#### Project File Options
+
+By default, `PublishSingleFile` bundles all files in the publish directory into the single executable. This behavior can be altered via the following configurations in the application's project file.
+
+* Exclude certain files in the publish directory from being bundled (ex: because of licensing issues or for testing purposes).
+
+```xml
+<ItemGroup>
+    <SingleFileExclude Include="List of patterns to exclude"/>
+</ItemGroup>
+```
+
+- Add additional files to the bundle from other locations (ex: add data files to the app). The developer will be responsible for handling these additional files.
+
+```xml
+<ItemGroup>
+    <SingleFileAdd Include="List of file paths"/>
+</ItemGroup>
+```
+
+An app may choose to have all of its dependencies extracted to disk at runtime (instead of loading certain files directly from the bundle) by setting the following property.
+
+```xml
+<PropertyGroup>
+    <SingleFileExtractAll>true</SingleFileExtractAll>
+</PropertyGroup>
+```
+
+#### VS Integration
+
+Developers should be able to use the feature easily from Visual Studio. The feature will start with text based support -- by explicitly setting the `PublishSingleFile` property in the project file. In future, we may provide a single-file publish-profile, or other UI triggers.
+
+## User Experience
+
+To summarize, here's the overall experience for creating a HelloWorld single-file app 
+
+*  Create a new HelloWorld app: `HelloWorld$ dotnet new console`
+
+#### Framework Dependent HelloWorld
+
+* Normal publish: `dotnet publish` 
+
+  * Publish directory contains the host `HelloWorld.exe` ,  the app `HelloWorld.dll`, configuration files `HelloWorld.deps.json`, `HelloWorld.runtimeconfig.json`, and the PDB-file `HelloWorld.pdb`.
+
+* Single-file publish: `dotnet publish /p:PublishSingleFile=true`
```

Comment:
Should PublishSingleFile change the default output directory of the publish? Incremental publish today already has some problems. I suspect publishing alternating between single-file and not will compound them if we're not careful. We might also take the opportunity to spec out the incrementality better, fix bugs, and handle this case carefully.

Reply:
Or we could not advertise using a global swtich on publish and instead recommend setting PublishSingleFile in project file based on configuration. In general, switching global properties between msbuild invocations is tricky business with respect to incrementality.

Reply:
The output directory issue is should be handled similarly for all post-processing tools (currently ILLinker, crossgen, and single-file). The publish process may temporarily use other directories, but ultimately, the publish directory should contain the single-file, I think.

I agree with the idea of marking the PublishSingleFile on the project file rather than the command line, in order to reduce confusion wrt incremental build.